pkg/operator/operator.go

```diff
@@ -134,108 +142,246 @@ func NewOperator(opts Options) Operator {
+	return func(obj interface{}) {
+		c, ok := obj.(*componentsapi.Component)
+		if ok {
+			log.Debugf("Observed component to be synced, %s/%s", c.Namespace, c.Name)
+			o.apiServer.OnComponentUpdated(ctx, c)
+		}
 	}
 }
 
-func (o *operator) loadCertChain(ctx context.Context) (certChain *credentials.CertChain) {
+func (o *operator) loadCertChain(ctx context.Context) (*credentials.CertChain, error) {
 	log.Info("Getting TLS certificates")
 
 	watchCtx, watchCancel := context.WithTimeout(ctx, time.Minute)
 	fsevent := make(chan struct{})
+	fserr := make(chan error)
+
 	go func() {
 		log.Infof("Starting watch for certs on filesystem: %s", o.config.Credentials.Path())
 		err := fswatcher.Watch(watchCtx, o.config.Credentials.Path(), fsevent)
 		// Watch always returns an error, which is context.Canceled if everything went well
-		if err != nil && !errors.Is(err, context.Canceled) {
-			log.Fatalf("Error starting watch on filesystem: %s", err)
+		if err != nil {
+			if !errors.Is(err, context.Canceled) {
+				// Ignore context.Canceled
+				fserr <- fmt.Errorf("Error starting watch on filesystem: %s", err)
+			} else {
+				fserr <- nil
+			}
+
+			return
 		}
+
 		close(fsevent)
-		if watchCtx.Err() == context.DeadlineExceeded {
-			log.Fatal("Timeout while waiting to load TLS certificates")
-		}
 	}()
 
+	var certChain *credentials.CertChain
 	for {
 		chain, err := credentials.LoadFromDisk(o.config.Credentials.RootCertPath(), o.config.Credentials.CertPath(), o.config.Credentials.KeyPath())
 		if err == nil {
 			log.Info("TLS certificates loaded successfully")
+			watchCancel()
 			certChain = chain
 			break
 		}
 		log.Infof("TLS certificate not found; waiting for disk changes. err=%v", err)
-		<-fsevent
-		log.Debug("Watcher found activity on filesystem")
+		select {
+		case <-fsevent:
+			log.Debug("Watcher found activity on filesystem")
+			continue
+		case <-watchCtx.Done():
+			return nil, errors.New("timeout while waiting to load TLS certificates")
+		}
 	}
 
-	watchCancel()
-
-	return certChain
+	return certChain, <-fserr
 }
 
-func (o *operator) Run(ctx context.Context) {
-	defer runtimeutil.HandleCrash()
-
+func (o *operator) Run(ctx context.Context) error {
 	log.Infof("Dapr Operator is starting")
+	healthzServer := health.NewServer(log)
 
-	go func() {
-		if err := o.mgr.Start(ctx); err != nil {
-			log.Fatalf("Failed to start controller manager, err: %s", err)
+	if err := o.mgr.Add(nonLeaderRunnable{func(ctx context.Context) error {
+		// start healthz server
+		if err := healthzServer.Run(ctx, healthzPort); err != nil {
+			return fmt.Errorf("Failed to start healthz server: %s", err)
 		}
-	}()
-	if !o.mgr.GetCache().WaitForCacheSync(ctx) {
-		log.Fatalf("Failed to wait for cache sync")
+		return nil
+	}}); err != nil {
+		return err
 	}
-	o.prepareConfig()
 
-	// load certs from disk
-	certChain := o.loadCertChain(ctx)
+	if err := o.mgr.Add(nonLeaderRunnable{func(ctx context.Context) error {
```

Comment:
Like this one where there's a function inside an if block, seems pretty hard to read :(